src/pkg/blob/dao/dao.go

```diff
@@ -162,13 +170,50 @@ func (d *dao) GetBlobByDigest(ctx context.Context, digest string) (*models.Blob,
 	return blob, nil
 }
 
+func (d *dao) UpdateBlobStatus(ctx context.Context, blob *models.Blob) (int64, error) {
+	o, err := orm.FromContext(ctx)
+	if err != nil {
+		return -1, err
+	}
+
+	// each update will auto increase version and update time
+	data := make(beego_orm.Params)
+	data["version"] = beego_orm.ColValue(beego_orm.ColAdd, 1)
+	data["update_time"] = time.Now()
+	data["status"] = blob.Status
+
+	qt := o.QueryTable(&models.Blob{})
+	cond := beego_orm.NewCondition()
+	var c *beego_orm.Condition
+
+	// In the multiple blob head scenario, if one request success mark the blob from StatusDelete to StatusNone, then version should increase one.
+	// in the meantime, the other requests tries to do the same thing, use 'where version >= blob.version' can handle it.
+	if blob.Status == models.StatusNone {
+		c = cond.And("version__gte", blob.Version)
+	} else {
+		c = cond.And("version", blob.Version)
+	}
+
+	/*
+		generated simple sql string.
+		UPDATE "blob" SET "version" = "version" + $1, "update_time" = $2, "status" = $3
+		WHERE "id" IN ( SELECT T0."id" FROM "blob" T0 WHERE T0."version" >= $4 AND T0."id" = $5 AND T0."status"  IN ('delete', 'deleting')  )
+	*/
+
+	return qt.SetCond(c).Filter("id", blob.ID).
+		Filter("status__in", models.StatusMap[blob.Status]).
```

Comment:
If the transition failed due to invalid state, nothing will be logged.

IMO it would be better to handle the transition at manager level, and no one should call the update blob bypassing the manager.

Reply:
log is added.